rdflib/namespace.py

```diff
@@ -163,55 +157,54 @@ def __new__(cls, value):
         return rt
 
     def __mod__(self, *args, **kwargs):
-        return URIRef(str(self).__mod__(*args, **kwargs))
+        return URIRef(super().__mod__(*args, **kwargs))
 
     def format(self, *args, **kwargs):
-        return URIRef(str.format(self, *args, **kwargs))
+        return URIRef(super().format(*args, **kwargs))
 
     def __repr__(self):
-        return "URIPattern(%r)" % str(self)
+        return f"URIPattern({self!r})"
 
 
-class ClosedNamespace(object):
+class ClosedNamespace(Namespace):
     """
     A namespace with a closed list of members
 
-    Trying to create terms not listen is an error
+    Trying to create terms not listed is an error
     """
 
-    def __init__(self, uri, terms):
-        self.uri = uri
-        self.__uris = {}
-        for t in terms:
-            self.__uris[t] = URIRef(self.uri + t)
+    def __new__(cls, uri, terms):
+        try:
+            rt = super().__new__(cls, uri)
+        except UnicodeDecodeError:
+            rt = super().__new__(cls, uri, "utf-8")
+        rt.__uris = {t: URIRef(rt + t) for t in terms}
+        return rt
+
+    @property
+    def uri(self):  # Back-compat
+        return str(self)
 
     def term(self, name):
         uri = self.__uris.get(name)
         if uri is None:
-            raise KeyError("term '{}' not in namespace '{}'".format(name, self.uri))
-        else:
-            return uri
+            raise KeyError(f"term '{name}' not in namespace '{self}'")
+        return uri
 
-    def __getitem__(self, key, default=None):
+    def __getitem__(self, key):
         return self.term(key)
 
     def __getattr__(self, name):
-        if name.startswith("__"):  # ignore any special Python names!
-            raise AttributeError
-        else:
-            try:
-                return self.term(name)
-            except KeyError as e:
-                raise AttributeError(e)
-
-    def __str__(self):
-        return str(self.uri)
+        try:
+            return self.term(name)
+        except KeyError as e:
+            raise AttributeError(e)
 
     def __repr__(self):
-        return "rdf.namespace.ClosedNamespace(%r)" % str(self.uri)
+        return f"{self.__module__}.{self.__class__.__name__}({self!r})"
 
     def __dir__(self):
-        return list(self._ClosedNamespace__uris)
```

Comment:
I guess this never worked, I vaguely recall hitting my head against this and being confused with the resulting error.